File: 카카오/2019카카오겨울인턴십_4.cpp
#include<bits/stdc++.h>
#include<unordered_map>

using namespace std;
using ll = long long;

unordered_map<ll, ll> uMap;

ll getRoom(ll num) {
    if (uMap[num] == 0) {
        uMap[num] = num + 1;
        return num;
    }
    return uMap[num] = getRoom(uMap[num]);
}

vector<ll> solution(ll k, vector<ll> room_number) {
    vector<ll> answer;
    for (ll num : room_number) {
        answer.push_back(getRoom(num));
    }
    return answer;
}

File: 카카오/2019카카오겨울인턴십_4.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Solution {
    Map<Long, Long> root = new HashMap<>();

    public long[] solution(long k, long[] room_number) {
        long[] answer = new long[room_number.length];
        for (int i = 0; i < room_number.length; i++) {
            answer[i] = getRoot(room_number[i]);
        }
        return answer;
    }

    private long getRoot(long x) {
        if (!root.containsKey(x)) {
            root.put(x, x + 1);
            return x;
        }

        root.put(x, getRoot(root.get(x)));
        return root.get(x);
    }
}